### PR TITLE
lib.meta.getLicenseFromSpdxIdOr: init

### PR DESCRIPTION
**Pull request by ShamrockLee**
###### Motivation for this change

This PR introduces `lib.meta.getLicenseFromSpdxIdOr`, a variant function of `lib.meta.getLicenseFromSpdxId` that allows the developers to specify the fallback value explicitly when a license with a matching SPDX ID is not found in `lib.licenses`.

Given that the typical use case of this function is to parse configuration files such as `package.json` and `pyproject.toml` during automatic package generation, the warning introduced by `lib.getLicenseFromSpdxId` will block the Release Tests and OfBorg evaluation for a programmatically generated package set.

`lib.getLicenseFromSpdxIdOr` enables developers to control the fallback behavior, leading to a smoother developer and user experience.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ShamrockLee** [comment]
@infinisil Do you have time taking a look at this?

**Comment by ShamrockLee** [comment]
Rebase and solve the merge conflict.

**Comment by ShamrockLee** [comment]
Bump!

File: lib/meta.nix [hunk 1]
```diff
@@ -139,18 +140,34 @@ rec {
...
+     lib.getLicenseFromSpdxId' "MIT" null == lib.licenses.mit
+     => true
+     lib.getLicenseFromSpdxId' "MY LICENSE" lib.licenses.free == lib.licenses.free
+     => true
+     lib.getLicenseFromSpdxId' "MY LICENSE" null
+     => null
+     lib.getLicenseFromSpdxId' "MY LICENSE" (builtins.abort "No SPDX ID matches MY LICENSE")
+     => error: evaluation aborted with the following error message: 'No SPDX ID matches MY LICENSE'
+  */
+  getLicenseFromSpdxId' =
```
Comment (infinisil) [inline lib/meta.nix:165]:
```suggestion
  getLicenseFromSpdxIdOr =
```

This is then similar to `spdxLicenses.${ ... } or default`

**Review by infinisil — CHANGES_REQUESTED** [review]
I have to agree that functions shouldn't throw a warning if you can't know whether the warning would be thrown without calling it.

If I could roll back time, I would've wanted `getLicenseFromSpdxId` to throw an error in case there's no match. Implicit defaults are not a great thing. I'd be interested in deprecating this function in the future, instead preferring `getLicenseFromSpdxIdOr`.

File: lib/meta.nix (hunk 1 above)
Comment (infinisil) [reply lib/meta.nix:165]: I'd say let's just introduce the -`Or` variant for now. One step at a time :)

Comment (ShamrockLee) [reply lib/meta.nix]: > I'd say let's just introduce the -Or variant for now. One step at a time :)

Addressed.

File: lib/meta.nix [hunk 2]
```diff
@@ -147,15 +146,36 @@ rec {
...
       );
 
+  /* Get the corresponding attribute in lib.licenses from the SPDX ID,
+     or fallback to the given default value.
+     For SPDX IDs, see https://spdx.org/licenses
+
+     Type:
+       getLicenseFromSpdxIdOr :: str -> Any -> Any
+
+     Example:
```
Comment (infinisil) [inline lib/meta.nix:162]: Could you add some tests in `lib/tests/misc.nix`? These examples would work well :)

**Review by infinisil — COMMENTED** [review]
Other than missing tests, this is looking good!

File: lib/meta.nix (hunk 2 above)
Comment (ShamrockLee) [reply lib/meta.nix:162]: Tests added for `getLicenseFromSpdxIdOr`

**Comment by ShamrockLee** [comment]
I added tests for `lib.meta.getLicenceFromSpdxIdOr` and modernized its documentation.

**Comment by ShamrockLee** [comment]
@infinisil, the test is added. Could you help review it again?

**Review by infinisil — APPROVED** [review]
Ah yes, sorry for the delay, great work!

**Pull request merged**